flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java

```diff
@@ -144,6 +149,116 @@ public static Operation convertChangeColumn(
         // TODO: handle watermark and constraints
     }
 
+    public static Operation convertRenameColumn(
+            ObjectIdentifier tableIdentifier,
+            String originColumnName,
+            String newColumnName,
+            CatalogTable catalogTable) {
+
+        Schema modifiedTableSchema = catalogTable.getUnresolvedSchema();
+        validateColumnName(originColumnName, newColumnName, modifiedTableSchema);
+
+        Schema.Builder builder = Schema.newBuilder();
+        // build column
+        modifiedTableSchema.getColumns().stream()
+                .forEach(
+                        column -> {
+                            if (StringUtils.equals(column.getName(), originColumnName)) {
+                                buildNewColumnFromOriginColumn(builder, column, newColumnName);
+                            } else {
+                                buildNewColumnFromOriginColumn(builder, column, column.getName());
+                            }
+                        });
+        // build primary key column
+        List<String> originPrimaryKeyNames =
+                modifiedTableSchema
+                        .getPrimaryKey()
+                        .map(Schema.UnresolvedPrimaryKey::getColumnNames)
+                        .orElseGet(Lists::newArrayList);
+
+        List<String> newPrimaryKeyNames =
+                originPrimaryKeyNames.stream()
+                        .map(
+                                pkName ->
+                                        StringUtils.equals(pkName, originColumnName)
+                                                ? newColumnName
+                                                : pkName)
+                        .collect(Collectors.toList());
+
+        if (newPrimaryKeyNames.size() > 0) {
+            builder.primaryKey(newPrimaryKeyNames);
+        }
+        // build watermark
+        modifiedTableSchema.getWatermarkSpecs().stream()
+                .forEach(
+                        watermarkSpec -> {
+                            String watermarkRefColumnName = watermarkSpec.getColumnName();
+                            Expression watermarkExpression = watermarkSpec.getWatermarkExpression();
+                            if (StringUtils.equals(watermarkRefColumnName, originColumnName)) {
+                                String newWatermarkExpression =
+                                        ((SqlCallExpression) watermarkExpression)
+                                                .getSqlExpression()
+                                                .replace(watermarkRefColumnName, newColumnName);
```

Comment:
1. We can't guarantee this is always `SqlCallExpression`. 
2. We can't use String `replace` for renaming, this is very error-prone, e.g. the original expressions is `f123 - f1` and `f1` is rename to `f2`, then the replaced result would be `f223 - f2` which is wrong.  

If we don't have a good way to replace column names in expression, we can disallow rename for columns referenced in expressions in the first version.

Reply:
OK. Computed expression and Watermark expression  disallow rename  in the first version?

Reply:
Would be better to investigate solutions first, I'm not sure whether there are any other ways to do this. 
cc @twalthr , do you have any ideas?

Reply:
No doubt that string replace is not a good solution. The expression can be both Table API or SQL. If we want to replace the column everywhere we would need to resolve the schema first and then traverse the nodes.

I would say we don't implement such a logic and fail hard in this case. An invalid schema will be detected when resolving the schema. Users can just provide a new computed column or watermark expression in the same ALTER statement.

Reply:
We can also through a helpful exception if the column is used for watermarks. For computed column, we cannot do much.

Reply:
Thanks @twalthr . I suggest resolve the schema first and traverse the node tree to check whether it contains the renamed column, and throw a helpful exception if contains.

Reply:
In any case, we should start putting each operation execution in a separate class. Otherwise `OperationConverterUtils` or `OperationConverter` grow infinitely. `OperationConverter` can just be a registry for mapping the instance to a class. We can then make certain entities available e.g. schema resolver or catalog manager.

Reply:
@wuchong  
Using org.apache.flink.table.catalog.DefaultSchemaResolver#resolve to resolve a Schema is  a good solution.
In my JIRA solution, SqlValidatorException can only throw Unknown identifier'xxx' because the expression has been validated when the table is created.

Reply:
Hi  @wuchong ,
traversing the resolved expression tree to see whether it contains renamed column is a better solution. the parsing process may be more complicated.

In my JIRA solution, SqlValidatorException can only throw Unknown identifier'xxx' because the resolved expression has been validated when the table is created.

Reply:
hi, @wuchong.
   If the calculated column expression contains a renamed column, it will throw 'Cannot resolve field [oldName], input field list:[xx, newName]' when resolving the schema.
  You said 'traverse the node tree to check whether it contains the renamed column', I don’t know where to add it. If use ResolverRule, how to distinguish it from ReferenceResolverRule.    
   thanks .

Reply:
> In any case, we should start putting each operation execution in a separate class. Otherwise `OperationConverterUtils` or `OperationConverter` grow infinitely. `OperationConverter` can just be a registry for mapping the instance to a class. We can then make certain entities available e.g. schema resolver or catalog manager.

@twalthr @wuchong In my limited read, it would be better if we would do the refactor on the parts of OperationConverter after completing FLINK-21634. Please let me know what's your opinion.